File: spring/src/main/java/com/example/userservice/freelancer/model/Freelancer.java
package com.example.userservice.freelancer.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import javax.persistence.Entity;
import java.util.Date;
import java.util.List;



@Data
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Document(collection = "freelancers")
public class Freelancer {

    @Id
    private String id;

    @Field("PhoneNumber")
    private Long phoneNumber;

    @Field("ProfilePicture")
    private String profilePicture;

    @Field("Email")
    private String email;

    @Field("Password")
    private String password;

    @Field("VerificationCode")
    private String verificationCode;

    @Field("AccountActivationStatus")
    private Boolean accountActivationStatus;

    @Field("AccountVerficiationStatus")
    private Boolean accountVerificationStatus;

    @Field("Earnings")
    private Double earnings;

    @Field("PayRate")
    private PayRate payRate;

    @Field("WorkHistory")
    private List<WorkHistory> workHistory;

    @Field("Schedule")
    private Schedule schedule;

    @Field("Messages")
    private List<String> messages;

    @Field("Languages")
    private List<String> languages;

    @Field("EstimateWorkLocation")
    private String estimateWorkLocation;

    @Field("WorkTitle")
    private WorkTitle workTitle;

    @Field("Speciality")
    private List<String> speciality;

    @Field("VerLinkExpDate")
    private Date verLinkExpDate;

    @Field("ProposedPrivateWorks")
    private List<ProposedPrivateWork> proposedPrivateWorks;

    // Constructors, getters, and setters

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class PayRate {
        @Field("HourlyRate")
        private Double hourlyRate;

        @Field("PayPerTaskRate")
        private Double payPerTaskRate;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class WorkHistory {
        @Field("Ongoing")
        private List<OngoingTask> ongoing;

        @Field("Finiched")
        private List<FinishedTask> finished;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class OngoingTask {
        @Field("TaskTitle")
        private String taskTitle;

        @Field("TaskHolder")
        private String taskHolder; // Assuming this is a string representation of ObjectId

        @Field("DueDate")
        private Date dueDate;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class FinishedTask {
        @Field("TaskTitle")
        private String taskTitle;

        @Field("TaskHolder")
        private String taskHolder; // Assuming this is a string representation of ObjectId

        @Field("EarningsMade")
        private Double earningsMade;

        @Field("Review")
        private String review;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Schedule {
        @Field("Weekly")
        private List<Date> weekly;

        @Field("Monthly")
        private List<Date> monthly;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class WorkTitle {
        @Field("WorkTitleId")
        private String workTitleId; // Assuming this is a string representation of ObjectId

        @Field("WorkTitleText")
        private String workTitleText;

        // Constructors, getters, and setters
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ProposedPrivateWork {
        @Field("PrivateJobOfferId")
        private String privateJobOfferId; // Assuming this is a string representation of ObjectId

        @Field("Status")
        private String status;

        // Constructors, getters, and setters
    }
}

File: TuniWorkMobile/app/src/main/java/com/example/myapplication/dataClasses/freelancer.kt
package com.example.myapplication.dataClasses

import com.google.gson.annotations.SerializedName
import java.util.Date

data class Freelancer(
    @SerializedName("Name") val name: String?,
    @SerializedName("Surname") val surname: String?,
    @SerializedName("PhoneNumber") val phoneNumber: Long?,
    @SerializedName("ProfilePicture") val profilePicture: String?,
    @SerializedName("Email") val email: String,
    @SerializedName("Password") val password: String,
    @SerializedName("VerificationCode") val verificationCode: String?,
    @SerializedName("AccountActivationStatus") val accountActivationStatus: Boolean = true,
    @SerializedName("AccountVerificationStatus") val accountVerificationStatus: Boolean = false,
    @SerializedName("Earnings") val earnings: Double = 0.0,
    @SerializedName("PayRate") val payRate: PayRate,
    @SerializedName("WorkHistory") val workHistory: WorkHistory,
    @SerializedName("Schedule") val schedule: List<String>,
    @SerializedName("Messages") val messages: List<String>,
    @SerializedName("Languages") val languages: List<String>,
    @SerializedName("EstimateWorkLocation") val estimateWorkLocation: EstimateWorkLocation,
    @SerializedName("WorkTitle") val workTitle: WorkTitle,
    @SerializedName("SavedWorkOffers") val savedWorkOffers: List<SavedWorkOffer>,
    @SerializedName("Speciality") val speciality: List<String>,
    @SerializedName("VerLinkExpDate") val verLinkExpDate: String?,
    @SerializedName("PassChangeLinkExpDate") val passChangeLinkExpDate: String?,
    @SerializedName("ProposedPrivateWorks") val proposedPrivateWorks: List<ProposedPrivateWork>,
    @SerializedName("PendingWorkOffers") val pendingWorkOffers: List<PendingWorkOffer>,
    @SerializedName("CompanyReceivedContracts") val companyReceivedContracts: List<String?>,
    @SerializedName("_id") val id: String,
    @SerializedName("Notifications") val notifications: List<Notification>,
)

data class PayRate(
    @SerializedName("HourlyRate") val hourlyRate: Double,
    @SerializedName("payPerTaskRate") val payPerTaskRate: Double
)
data class Notification(
    @SerializedName("NotificationMessage") val notificationMessage: String,
    @SerializedName("readStatus") val readStatus: Boolean,
    @SerializedName("senderInformations") val senderInformation: SenderInformations, // Fixed the property name
    @SerializedName("_id") val id: String
)

data class SenderInformations( // Renamed the class to match the corrected property name
    @SerializedName("senderId") val senderId: String,
    @SerializedName("senderUserType") val senderUserType: String,
    @SerializedName("creationDate") val creationDate: Date,
    @SerializedName("context") val context: String,
    @SerializedName("objectId") val objectId: String
)

data class WorkHistory(
    @SerializedName("Ongoing") val ongoing: List<OngoingTask>,
    @SerializedName("Finished") val finished: List<FinishedTask>
)

data class OngoingTask(
    @SerializedName("TaskTitle") val taskTitle: String?,
    @SerializedName("TaskHolder") val taskHolder: String,
    @SerializedName("DueDate") val dueDate: String?
)

data class FinishedTask(
    @SerializedName("TaskTitle") val taskTitle: String?,
    @SerializedName("TaskHolder") val taskHolder: String,
    @SerializedName("EarningsMade") val earningsMade: Double?,
    @SerializedName("Review") val review: String?
)

data class Schedule(
    @SerializedName("Weekly") val weekly: List<String>,
    @SerializedName("Monthly") val monthly: List<String>
)

data class EstimateWorkLocation(
    @SerializedName("City") val city: String?,
    @SerializedName("Municipality") val municipality: String?
)

data class WorkTitle(
    @SerializedName("WorkTitleId") val workTitleId: String,
    @SerializedName("WorkTitleText") val workTitleText: String?
)

data class SavedWorkOffer(
    @SerializedName("WorkId") val workId: String,
    @SerializedName("WorkTitle") val workTitle: String?,
    @SerializedName("WorkDescription") val workDescription: String?
)

data class ProposedPrivateWork(
    @SerializedName("PrivateJobOfferId") val privateJobOfferId: String,
    @SerializedName("Status") val status: String = "awaiting response"
)

data class PendingWorkOffer(
    @SerializedName("PublicJobOfferId") val publicJobOfferId: String,
    @SerializedName("Status") val status: String = "awaiting company response",
    @SerializedName("PWOInfos") val pwoInfos: PWOInfos
)

data class PWOInfos(
    @SerializedName("CName") val cName: String?,
    @SerializedName("TitlePWO") val titlePWO: String?,
    @SerializedName("DescriptionPWO") val descriptionPWO: String?
)